### PR TITLE
Properly reflect control flow for object/dynamic logical operators in control flow graph.

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@333fred, @mavasani, @cston, @dotnet/roslyn-compiler, @dotnet/analyzer-ioperation Please review

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-infrastructure It looks like CI jobs are not making any progress

**Comment by AlekseyTs** [comment]
@333fred, @mavasani, @cston, @dotnet/roslyn-compiler, @dotnet/analyzer-ioperation Please review

**Review by cston — APPROVED** [review]

**Comment by AlekseyTs** [comment]
@333fred, @dotnet/roslyn-compiler Please review, need second review

**Review by mavasani — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Is it possible for there to be default parameters, even in an error scenario?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: No, signature of operators is fixed.

---
In reply to: [191545921](https://github.com/dotnet/roslyn/pull/27155#discussion_r191545921) [](ancestors = 191545921)

**Review by 333fred — APPROVED** [review]
LGTM (Iteration 1)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (jcouv) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >binOp.ConstantValue [](start = 90, length = 19)

Is it possible to have a constant value here? 
If so, wouldn't that mean that either the left or the right had a constant value, which should be reflected in the conversions above (line 1746 and 1759)?

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: > Is it possible to have a constant value here? 

No, I do not think an expression of type Object can have a boolean constant value.

---
In reply to: [191549014](https://github.com/dotnet/roslyn/pull/27155#discussion_r191549014) [](ancestors = 191549014)

Comment (jcouv) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Consider using `constantValue: default` then. Thanks

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: > Consider using constantValue: default then. 

We prefer using information from original nodes rather than making guesses in some particular cases, but not in others.

---
In reply to: [191581103](https://github.com/dotnet/roslyn/pull/27155#discussion_r191581103) [](ancestors = 191581103)

**Pull request merged**
